### PR TITLE
new(arrow/compute): temporal rounding methods

### PR DESCRIPTION
**Pull request by hamilton-earthscope**
### Rationale for this change

Apache Arrow Go lacks temporal rounding functionality available in PyArrow (`floor_temporal`, `ceil_temporal`, `round_temporal`) - [reference](https://arrow.apache.org/docs/python/generated/pyarrow.compute.floor_temporal.html). This prevents users from rounding timestamp or date values to calendar units (year, quarter, month, week, day) and sub-day units (hour, minute, second, etc.), which is essential for time-series data aggregation and analysis.

### What changes are included in this PR?

- **New compute functions**: `FloorTemporal`, `CeilTemporal`, `RoundTemporal` with support for:
  - Calendar units: year, quarter, month, week, day
  - Sub-day units: hour, minute, second, millisecond, microsecond, nanosecond
  - Multiple rounding: round to N units (e.g., 2-week, 3-month periods)
  - Rounding modes: floor, ceil, half-up, half-down, half-to-even
  - Timezone-aware rounding: respects DST and local date boundaries
  - Calendar-based origin: round relative to start of day for sub-day units

- **Implementation details**:
  - `RoundTemporalOptions` for configuring unit, multiple, week start, strict ceiling, and calendar origin
  - Timezone-aware calendar arithmetic for variable-length periods
  - Support for timezone-naive and timezone-aware timestamps
  - Support for `timestamp`, `date32`, `date64`, `time32` and `time64`
  - PyArrow's behavior was consulted for edge cases
    - rounding semantics (i.e. calendar rounding happens in specified timezone)
    - negative values

### Are these changes tested?

Yes, comprehensive test coverage includes:
- test cases covering most(?) types, units, modes, and edge cases
  - Timezone-aware tests
  - Multi-period rounding tests
  - Half-rounding mode tests
  - Null handling and error validation tests
  - `timestamp`, `date32`, `date64`, `time32` and `time64` tests
- Performance benchmarks

### Are there any user-facing changes?

**New public API**:
```go
// Functions
compute.FloorTemporal(arr, opts)
compute.CeilTemporal(arr, opts)
compute.RoundTemporal(arr, opts)

// Options
compute.RoundTemporalOptions{
    Multiple: 1,
    Unit: compute.RoundTemporalMonth,
    WeekStartsMonday: true,
    CeilIsStrictlyGreater: false,
    CalendarBasedOrigin: false,
}

// Units
compute.RoundTemporalYear, RoundTemporalQuarter, RoundTemporalMonth,
RoundTemporalWeek, RoundTemporalDay, RoundTemporalHour,
RoundTemporalMinute, RoundTemporalSecond, RoundTemporalMillisecond,
RoundTemporalMicrosecond, RoundTemporalNanosecond
```

### CODE REVIEW

File: arrow/compute/internal/kernels/rounding.go [hunk 1]
```diff
@@ -807,3 +808,681 @@ func FixedRoundDecimalExec[T decimal128.Num | decimal256.Num](mode RoundMode) ex
...
+func unitScaleFactor(unit arrow.TimeUnit) int64 {
+	switch unit {
+	case arrow.Second:
+		return 1_000_000_000
+	case arrow.Millisecond:
+		return 1_000_000
+	case arrow.Microsecond:
+		return 1_000
+	case arrow.Nanosecond:
+		return 1
```
Comment (zeroshade) [inline arrow/compute/internal/kernels/rounding.go]: `int64(unit.Multiplier())` already does this, no need for this function

File: arrow/compute/internal/kernels/rounding.go [hunk 2]
```diff
@@ -807,3 +808,681 @@ func FixedRoundDecimalExec[T decimal128.Num | decimal256.Num](mode RoundMode) ex
...
+
+// canRoundInInputUnit checks if rounding can be done in the input unit
+// without converting to nanoseconds (true when rounding interval is evenly divisible).
+func canRoundInInputUnit(inputUnit arrow.TimeUnit, roundingIntervalNanos int64) bool {
+	return roundingIntervalNanos%unitScaleFactor(inputUnit) == 0
+}
+
+// convertToNanos converts a timestamp value to nanoseconds
+func convertToNanos(ts int64, unit arrow.TimeUnit) int64 {
+	return ts * unitScaleFactor(unit)
```
Comment (zeroshade) [inline arrow/compute/internal/kernels/rounding.go]:
```suggestion
	return ts * int64(unit.Multiplier())
```

File: arrow/compute/internal/kernels/rounding.go [hunk 3]
```diff
@@ -807,3 +808,681 @@ func FixedRoundDecimalExec[T decimal128.Num | decimal256.Num](mode RoundMode) ex
...
+}
+
+// convertFromNanos converts a nanosecond timestamp to the specified unit
+func convertFromNanos(nanos int64, unit arrow.TimeUnit) int64 {
+	return nanos / unitScaleFactor(unit)
```
Comment (zeroshade) [inline arrow/compute/internal/kernels/rounding.go]:
```suggestion
	return nanos / int64(unit.Multiplier())
```

File: arrow/compute/internal/kernels/rounding.go [hunk 4]
```diff
@@ -807,3 +808,681 @@ func FixedRoundDecimalExec[T decimal128.Num | decimal256.Num](mode RoundMode) ex
...
+func (m *timestampUnitMatcher) String() string {
+	return "timestamp(unit=" + m.unit.String() + ")"
+}
+
+func (m *timestampUnitMatcher) Equals(other exec.TypeMatcher) bool {
+	if o, ok := other.(*timestampUnitMatcher); ok {
+		return m.unit == o.unit
+	}
+	return false
+}
```
Comment (zeroshade) [inline arrow/compute/internal/kernels/rounding.go]: we already have a matcher `TimestampTypeUnit` in `exec/kernel.go` that you can use instead of this

File: arrow/compute/internal/kernels/rounding.go [hunk 5]
```diff
@@ -807,3 +808,681 @@ func FixedRoundDecimalExec[T decimal128.Num | decimal256.Num](mode RoundMode) ex
...
+	}
+	return "date64"
+}
+
+func (m *dateTypeMatcher) Equals(other exec.TypeMatcher) bool {
+	if o, ok := other.(*dateTypeMatcher); ok {
+		return m.dateTypeID == o.dateTypeID
+	}
+	return false
+}
```
Comment (zeroshade) [inline arrow/compute/internal/kernels/rounding.go]: can you move this to `exec/kernel.go` and expose it with a function like the other matchers we have?

File: arrow/compute/internal/kernels/rounding.go [hunk 6]
```diff
@@ -807,3 +808,681 @@ func FixedRoundDecimalExec[T decimal128.Num | decimal256.Num](mode RoundMode) ex
...
+	}
+	return fmt.Sprintf("time64[%s]", m.unit)
+}
+
+func (m *timeTypeMatcher) Equals(other exec.TypeMatcher) bool {
+	if o, ok := other.(*timeTypeMatcher); ok {
+		return m.timeTypeID == o.timeTypeID && m.unit == o.unit
+	}
+	return false
+}
```
Comment (zeroshade) [inline arrow/compute/internal/kernels/rounding.go]: This already exists via `Time32TypeUnit(....)` and `Time64TypeUnit(...)` etc. in `exec/kernel.go`

**Review by zeroshade — CHANGES_REQUESTED** [review]
